File: tests/code/corpus/langs/go/shapes.go
package geometry

import (
	"fmt"
	"math"
)

type Shape interface {
	Area() float64
}

type Circle struct {
	Radius float64
}

func (c Circle) Area() float64 {
	return math.Pi * c.Radius * c.Radius
}

func (c Circle) Describe() string {
	return fmt.Sprintf("circle of area %f", c.Area())
}

func MakeCircle(radius float64) Circle {
	return Circle{Radius: radius}
}

func TotalArea(shapes []Shape) float64 {
	total := 0.0
	for _, shape := range shapes {
		total += shape.Area()
	}
	return total
}

File: tests/code/corpus/langs/lua/shapes.lua
local math = require("math")

local Shape = {}

function Shape.area(self)
  return 0.0
end

function Shape:describe()
  return tostring(self:area())
end

local Circle = {}

function Circle.area(self)
  return math.pi * self.radius * self.radius
end

local function makeCircle(radius)
  return { radius = radius }
end

function totalArea(shapes)
  local total = 0.0
  for _, shape in ipairs(shapes) do
    total = total + Circle.area(shape)
  end
  return total
end
